backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/TestingSession.java

```diff
@@ -0,0 +1,245 @@
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.backends.cassandra;
+
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiFunction;
+import java.util.function.Predicate;
+
+import com.datastax.driver.core.BoundStatement;
+import com.datastax.driver.core.CloseFuture;
+import com.datastax.driver.core.Cluster;
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.RegularStatement;
+import com.datastax.driver.core.ResultSet;
+import com.datastax.driver.core.ResultSetFuture;
+import com.datastax.driver.core.Session;
+import com.datastax.driver.core.Statement;
+import com.google.common.util.concurrent.ListenableFuture;
+
+public class TestingSession implements Session {
+    enum Behavior {
+        THROW((session, statement) -> {
+            RuntimeException injected_failure = new RuntimeException("Injected failure");
+            injected_failure.printStackTrace();
+            throw injected_failure;
+        }),
+        EXECUTE_NORMALLY(Session::executeAsync);
+
+        private final BiFunction<Session, Statement, ResultSetFuture> behaviour;
+
+        Behavior(BiFunction<Session, Statement, ResultSetFuture> behaviour) {
+            this.behaviour = behaviour;
+        }
+
+        ResultSetFuture execute(Session session, Statement statement) {
+            return behaviour.apply(session, statement);
+        }
+    }
+
+    @FunctionalInterface
+    interface StatementPredicate extends Predicate<Statement> {
+
+    }
+
+    static class BoundStatementStartingWith implements StatementPredicate {
+        private final String queryStringPrefix;
+
+        BoundStatementStartingWith(String queryStringPrefix) {
+            this.queryStringPrefix = queryStringPrefix;
+        }
+
+        @Override
+        public boolean test(Statement statement) {
+            if (statement instanceof BoundStatement) {
+                BoundStatement boundStatement = (BoundStatement) statement;
+                return boundStatement.preparedStatement()
+                    .getQueryString()
+                    .startsWith(queryStringPrefix);
+            }
+            return false;
+        }
+    }
+
+    @FunctionalInterface
+    public interface RequiresCondition {
+        RequiresApplyCount condition(StatementPredicate statementPredicate);
+
+        default RequiresApplyCount always() {
+            return condition(ALL_STATEMENTS);
+        }
+
+        default RequiresApplyCount whenBoundStatementStartsWith(String queryStringPrefix) {
+            return condition(new BoundStatementStartingWith(queryStringPrefix));
+        }
+    }
+
+    @FunctionalInterface
+    public interface RequiresApplyCount {
+        FinalStage times(int applyCount);
```

Comment:
I don't understand how to use this API without reading the tests. Could we make it clear that `times` apply to the failure condition stated above? what about "fail after that many calls" (I think it's a very common case)